library/src/main/scala/scala/scalajs/runtime/RuntimeLong.scala

```diff
@@ -55,531 +121,558 @@ final class RuntimeLong(
+    else bhi < ahi // workaround here
+  }
 
+  def >=(b: RuntimeLong): Boolean = {
+    /* Work around https://code.google.com/p/v8/issues/detail?id=3304
+     * 0x7fffffff > 0x80000000 is broken, so use < instead.
+     * This happens when comparing MaxValue to MinValue.
+     */
+    val ahi = a.hi
+    val bhi = b.hi
+    if (ahi == bhi) inlineUnsignedInt_>=(a.lo, b.lo)
+    else bhi < ahi // workaround here
   }
 
-  /**
-   * logical right shift
-   */
-  def >>>(n_in: Int): RuntimeLong = {
-    val n = n_in & 63
-    if (n < BITS) {
-      val remBits = BITS - n
-      masked((x.l >> n) | (x.m << remBits),
-             // FIXME is this really >> and not >>>??
-             (x.m >> n) | (x.h << remBits),
-             x.h >>> n)
-    } else if (n < BITS01) {
-      val shfBits = n - BITS
-      val remBits = BITS01 - n
-             // FIXME is this really >> and not >>>??
-      masked((x.m >> shfBits) | (x.h << remBits),
-             x.h >>> shfBits, 0)
-    } else {
-      masked(x.h >>> (n - BITS01), 0, 0)
-    }
+  // Bitwise operations
+
+  def unary_~ : RuntimeLong = // scalastyle:ignore
+    new RuntimeLong(~lo, ~hi)
+
+  def |(b: RuntimeLong): RuntimeLong =
+    new RuntimeLong(a.lo | b.lo, a.hi | b.hi)
+
+  def &(b: RuntimeLong): RuntimeLong =
+    new RuntimeLong(a.lo & b.lo, a.hi & b.hi)
+
+  def ^(b: RuntimeLong): RuntimeLong =
+    new RuntimeLong(a.lo ^ b.lo, a.hi ^ b.hi)
+
+  // Shifts
+
+  /** Shift left */
+  def <<(n0: Int): RuntimeLong = {
+    val n = n0 & 63
+    val lo = this.lo
+
+    if (n == 0) this
+    else if (n < 32) new RuntimeLong(lo << n, (lo >>> -n) | (hi << n))
+    else new RuntimeLong(0, lo << n)
   }
 
-  /**
-   * arithmetic right shift
-   */
-  def >>(n_in: Int): RuntimeLong = {
-    val n = n_in & 63;
-
-    // Sign extend x.h
-    val negative = (x.h & SIGN_BIT_VALUE) != 0
-    val xh = if (negative) x.h | ~MASK_2 else x.h
-
-    if (n < BITS) {
-      val remBits = BITS - n
-      // FIXME IMHO the first two >> should be >>>
-      masked((x.l >> n) | (x.m << remBits),
-             (x.m >> n) | (xh  << remBits),
-             xh >> n)
-    } else if (n < BITS01) {
-      val shfBits = n - BITS
-      val remBits = BITS01 - n
-      // FIXME IMHO the first >> should be >>>
-      masked((x.m >> shfBits) | (xh << remBits),
-              xh  >> shfBits,
-             if (negative) MASK_2 else 0)
-    } else {
-      masked(xh >> (n - BITS01),
-             if (negative) MASK   else 0,
-             if (negative) MASK_2 else 0)
-    }
+  @inline
+  private def inline_<<(lo: Int, hi: Int, n: Int): (Int, Int) = {
+    if (n == 0) (lo, hi)
+    else if (n < 32) (lo << n, (lo >>> -n) | (hi << n))
+    else (0, lo << n)
+  }
+
+  /** Logical shift right */
+  def >>>(n0: Int): RuntimeLong = {
+    val n = n0 & 63
+    val hi = this.hi
 
+    if (n == 0) this
+    else if (n < 32) new RuntimeLong((lo >>> n) | (hi << -n), hi >>> n)
+    else new RuntimeLong(hi >>> n, 0)
```

Comment:
There is `this`, but most importantly it's the multi-path result of a tuple which is not that great. Better to push the `new RuntimeLong`s inside the branches of the if/else.